### ISSUE TITLE
Recursive homomorphic mapped types do not work with array properties

### ISSUE THREAD

**Issue by httpdigest** [microsoft/TypeScript#17274]
**TypeScript Version:**  2.4.2-insiders.20170603 and 2.4.1 npm release

**Code:** 1.15.0-insiders (commit 3cfa565abb)

I've just recently discovered recursive homomorphic mapped types, and it is a very very useful feature to implement ES2015 Proxies. Thank you for that!
I have a simple Proxy which augments each accessed property via a Function type. However, array properties do not seem to work.

Given the following interface type definitions and the `pathOf()` function at the end...
```TypeScript
interface B {
  c?: number;
}
interface A {
  b?: B[];
  c?: {
    d?: number;
  }
}
type ObjectPath = ...;
type ObjectPathBuilder<T> = (() => ObjectPath) & {
    [P in keyof T]: ObjectPathBuilder<T[P]>
};
function pathOf<T>(_obj: T): ObjectPathBuilder<T> {
    return undefined; // <- actually it did return a ES2015 Proxy instance
}
```
...the following computes the correct types of intermediate expressions and the complete type for `res1` and `res2`...:
```TypeScript
const a: A = undefined;
// Below I can call c() as a function and also see 'd' as property:
const res1 = pathOf(a).c();
const res2 = pathOf(a).c.d();
```
...however, the following gives a `[ts] Element implicitly has an 'any' type because type 'ObjectPathBuilder<B[]>' has no index signature.` error message when I try to index into b:
```TypeScript
const a: A = undefined;
const res3 = pathOf(a).b[0](); // <- compiler error
const res4 = pathOf(a).b[0].c(); // <- same compiler error
```
**Expected behavior:**
Correct typing of the expression `pathOf(a).b` as "something with an indexing operator"

**Actual behavior:**
[ts] Element implicitly has an 'any' type because type 'ObjectPathBuilder<B[]>' has no index signature.

**Comment by RyanCavanaugh** [comment]
Can't check this without `ObjectPath` - please log a new issue with repro if needed. Thanks!

**Comment by httpdigest** [comment]
Take for example `type ObjectPath = string;`.
However, after two years, the issue seems to have been resolved already at least in TypeScript `3.5.2`.
Thanks for getting back, though!
